method comprision docs

ved1beta opened the pull request:
# docs: Add method comparison documentation

This PR adds method comparison docs for PEFT methods, split from the original PR #2504 . The changes include:

1. Added a new section in developer guides for method comparison
2. Created a main method comparison guide (`method_comparison.md`) that serves as an entry point
3. Added detailed documentation for three methods:
   - LoRA (Low-Rank Adaptation)
   - LoRA-FA (LoRA with Fast Adaptation)
   - Bone (Bottleneck Orthogonal Network)
4. Updated the table of contents to include the new documentation section
This is the first part of the original PR, focusing on documentation improvements. Future PRs will address the method comparison suite enhancements and testing components.

Also please let me know hoe to adders dummy values thanks : )

BenjaminBossan commented on docs/source/developer_guides/method_comparison/bone.md:
````diff
@@ -0,0 +1,195 @@
+)
+```
+
+## Hyperparameter Tuning
````
How did you derive the values in this section?

BenjaminBossan commented on docs/source/developer_guides/method_comparison/bone.md:
```diff
@@ -0,0 +1,195 @@
+|--------|------------------|----------------|-----------------------|
+| Bone   | Very High       | Fast          | Small-Medium         |
+| LoRA   | High            | Fast          | All                  |
+| Adapter | Medium         | Medium        | All                  |
+| Prompt | Very High      | Very Fast     | All                  |
```
What are "Adapter" and "Prompt"?

ved1beta replied:
resolved

BenjaminBossan commented on docs/source/developer_guides/method_comparison/bone.md:
```diff
@@ -0,0 +1,195 @@
+
+3. **Training Tips**
+   - Use learning rate 5e-5 to 2e-4
+   - Monitor orthogonal condition
```
How would you do that?

BenjaminBossan commented on docs/source/developer_guides/method_comparison/bone.md:
````diff
@@ -0,0 +1,195 @@
+)
+```
+
+## Examples
````
I don't think we need a section for "Examples" and also "Basic Usage". As mentioned above, let's link to the existing examples instead. If those are insufficient, I would much rather like to see those examples extended, e.g. to show a text classification task.

BenjaminBossan commented on docs/source/developer_guides/method_comparison/lora.md:
```diff
@@ -0,0 +1,196 @@
+- Memory efficient (~0.1% of base model parameters)
+- Minimal impact on inference speed
+- Easy to implement and use
+- Compatible with most transformer architectures
```
It's worth mentioning that LoRA is the most feature rich PEFT method, especially when it comes to supporting multiple quantization methods.

BenjaminBossan commented on docs/source/developer_guides/method_comparison/lora.md:
For LoRA and LoRA-FA, many of the same comments apply as for Bone, like how the numbers were derived, and to avoid duplications.

BenjaminBossan commented on docs/source/developer_guides/method_comparison/bone.md:
```diff
@@ -0,0 +1,195 @@
+| Metric | Value |
+|--------|-------|
+| Training Speed | Fast |
+| Convergence | Quick (typically 1-2 epochs) |
```
Can we really give general statements like this?

BenjaminBossan commented on docs/source/developer_guides/method_comparison/lora.md:
```diff
@@ -0,0 +1,196 @@
+### Not Recommended For
+- Tasks requiring extensive model modifications
+- Very small models (< 100M parameters)
+- Real-time applications with strict latency requirements
```
Although, at least for the case of serving a single LoRA adapter, we can just `merge_and_unload` to eliminate any LoRA overhead.

BenjaminBossan commented on docs/source/developer_guides/method_comparison/lora.md:
```diff
@@ -0,0 +1,196 @@
+| Parameter | Recommended Range | Impact |
+|-----------|------------------|--------|
+| rank (r) | 4-32 | Higher = better performance, more parameters |
+| alpha | 8-64 | Controls scaling of LoRA weights |
```
Let's mention that `alpha==2*rank` is a good starting point, as well as `use_rslora=True`.

BenjaminBossan commented on docs/source/developer_guides/method_comparison/lora.md:
```diff
@@ -0,0 +1,196 @@
+   - Consider model architecture
+
+3. **Training Tips**
+   - Use learning rate 1e-4 to 5e-4
```
IIRC, some papers recommend using 10x the learning rate compared to full fine-tuning when using LoRA.

BenjaminBossan commented on docs/source/developer_guides/method_comparison/lora.md:
```diff
@@ -0,0 +1,196 @@
+    r=8,
+    lora_alpha=32,
+    target_modules=["q_proj", "v_proj"],  # Focus on key layers
+    lora_dropout=0.0,  # Remove dropout for speed
```
Is this a huge factor (for DoRA, it is, but for normal LoRA too?)

BenjaminBossan commented on docs/source/developer_guides/method_comparison/lora.md:
````diff
@@ -0,0 +1,196 @@
+```python
+# Reduce memory usage
+config = LoraConfig(
+    r=4,  # Lower rank
````
Generally, the rank will only have a small impact on overall memory usage. It's not wrong to write this, but IMO a typical user should try other steps first (quantized models, batch size, activation checkpointing, etc.)

BenjaminBossan commented on docs/source/developer_guides/method_comparison/bone.md:
```diff
@@ -0,0 +1,195 @@
+    bottleneck_size=64,
+    target_modules=["attention.output"],
+    dropout=0.1,
+    use_orthogonal=True,
```
Let's pass the appropriate `task_type`.

BenjaminBossan commented on docs/source/developer_guides/method_comparison/lora_fa.md:
```diff
@@ -0,0 +1,203 @@
+### Not Recommended For
+- Tasks requiring extensive model modifications
+- Very small models (< 100M parameters)
+- Non-AdamW optimizers
```
"Not Recommended For" non-AdamW is not putting it quite correctly, in fact it is only implemented for AdamW, so other optimizers are not supported.

BenjaminBossan commented on docs/source/developer_guides/method_comparison/lora_fa.md:
```diff
@@ -0,0 +1,203 @@
+
+# Define LoRA-FA configuration
+config = LoraConfig(
+    r=16,  # higher rank recommended for LoRA-FA
```
```suggestion
    r=16,  # higher rank recommended for LoRA-FA compared to LoRA
```

BenjaminBossan commented on docs/source/developer_guides/method_comparison/lora_fa.md:
```diff
@@ -0,0 +1,203 @@
+    target_modules=["q_proj", "v_proj"],
+    lora_dropout=0.05,
+    bias="none",
+    use_fast_adapter=True,  # Enable LoRA-FA
```
This option does not exist, how did you come to this?

BenjaminBossan commented on docs/source/developer_guides/method_comparison/lora_fa.md:
```diff
@@ -0,0 +1,203 @@
+    target_modules=["q_proj", "k_proj", "v_proj", "o_proj"],
+    lora_dropout=0.1,
+    bias="lora_only",
+    use_fast_adapter=True,
+    fast_adapter_rank=8,  # specific rank for fast adaptation
```
These options do not exist, how did you come to this?

ved1beta replied:
removed

BenjaminBossan reviewed (changes requested):
Thanks @ved1beta for splitting up the original PR, this makes it much easier to review the changes.

First of all, I want to thank you for the effort you put into this. I've added a lot of comments but I think overall, this could add great value to PEFT users. Note that I have added some comments only to one section but they would apply equally to other sections.

At least for some passages, I got the impression that they were written with the help of LLMs. This is not a problem by itself, but please ensure to double check the correctness of what was written.

Moreover, [building the docs currently fails](https://github.com/huggingface/peft/actions/runs/14609490069/job/41009980783?pr=2509#step:15:59) because the referenced markdown files are not found. You can run the doc build locally as described [here](https://github.com/huggingface/peft/tree/main/docs#building-the-documentation) to ensure that there are no errors.

ved1beta commented:
I'll address all the points you've raised:
The documentation gaps you've identified are correct , and I'll work on filling those gaps systematically, and i agree there are too many LLM hallucinations will run the scripts and get the relevant data numbers and submit an updated version addressing these changes shortly.

Thanks again for your valuable feedback!

ved1beta commented:
changes include added : added relevant number in the tables , removed wrong and bold claims i have addressed most of the changes asked for alsoo need to fix docs build resolving the conversations i think are resolved

ved1beta commented on docs/source/developer_guides/method_comparison/bone.md:
```diff
@@ -0,0 +1,195 @@
+# Bone (Bottleneck Orthogonal Network)
+
+## Overview
+Bone is a parameter-efficient fine-tuning method that uses orthogonal transformations in bottleneck layers. It's particularly effective for small to medium-sized models and offers excellent memory efficiency.
```
removing this

ved1beta commented on docs/source/developer_guides/method_comparison/bone.md:
```diff
@@ -0,0 +1,195 @@
+
+## Implementation
+
+### Basic Usage
```
on it

BenjaminBossan reviewed (changes requested):
I did not do a complete in-depth review, but have added some comments. I think it is important that we:

- avoid duplicating information that can already be found elsewhere
- provide transparency on how the numbers and conclusions were derived

I assume that many of the numbers you present stem from the PR that adds the scripts. Therefore, I think we should prioritize that PR. Just as an example, if we find an error there that changes the results, we would need to update the numbers in this PR.

> i will need some help with these , what code examples to add in files

As mentioned, let's link to existing examples, e.g.:

- Bone: https://github.com/huggingface/peft/tree/main/examples/bone_finetuning
- LoRA-FA: https://github.com/huggingface/peft/tree/main/examples/lorafa_finetune
- LoRA: Lots of choices, just one example https://github.com/huggingface/peft/blob/main/examples/token_classification/peft_lora_ner.ipynb

githubnemo commented:
*not stale*
